Add Interactivity API runtime

DAreRodz opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

Adds the Interactivity API runtime to the `block-library` package.

Tracking issue: https://github.com/WordPress/gutenberg/issues/49868

## Why?

To provide a basic runtime in order to start experimenting with the Interactivity API in some of the core blocks.

## How?

- Moving the Interactivity API runtime modules from [this branch](https://github.com/WordPress/block-interactivity-experiments/tree/refactoring-for-gutenberg) to `/block-library/src/utils` folder.
- Adding a separate webpack config for blocks using the Interactivity API.
- Registering the Interactivity API runtime bundles so interactive blocks can enqueue them as dependencies.
- Including the `defer` attribute in all Interactivity API scripts.
- Including the `interactivity/index.js` file in `sideEffects`, as it's the one that inits the Interactivity API runtime (I forgot to mention it in the video 😅).

Here's a video explaining the changes done and how to start using the Interactivity API in core blocks.

<a href="https://www.loom.com/share/e295d9eccbc6496e8f6d55f1b878f795">
    <img style="max-width:200px;" src="https://cdn.loom.com/sessions/thumbnails/e295d9eccbc6496e8f6d55f1b878f795-with-play.gif">
  </a>
<a href="https://www.loom.com/share/e295d9eccbc6496e8f6d55f1b878f795">
    <p>Add Interactivity API runtime by DAreRodz · Pull Request #49994 · WordPress/gutenberg - 21 April 2023 - Watch Video</p>
  </a>

luisherranz committed:
Add interactivity runtime

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Add it to the image block

This is still pretty basic, just to check that it works.

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Add a separate webpack config

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Make sure the runtime is imported only once

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Use sideEffects instead of init

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Move script registration to a general file

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Add `defer` to the interactivity scripts

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Revert changes of the image block

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Fix init import name

Co-authored-by: David Arenas <david.arenas@automattic.com>

dmsnell commented:
Curious on including this in the `block-library` package. Is this something that's more or less necessary or does it make sense to create a new package where can separate the interactivity API from the block library package?

The block library can import from other packages, and I think having these in separate directories has some clear value too.

luisherranz commented:
Yes, but not yet. This preliminary version is only meant to be used internally by core blocks. We don't even want to expose it yet as experimental.

gziolo commented on tools/webpack/blocks.js:
```diff
@@ -74,134 +74,209 @@ const createEntrypoints = () => {
+	},
+	{
+		entry: {
+			// blockname: './packages/block-library/src/blockname/interactivity.js',
```
How do you envision adding new entry points whenever a core block gets new view script that uses Interactivity API?

DAreRodz replied:
For now, the idea is to do that manually, as the Interactivity API won't be available for developers yet. We can just add a new entry point for each of the core blocks that will use the Interactivity API.

That will of course change once we allow developers to experiment with the new API. 🙂

gziolo replied:
Sounds good 👍🏻

gziolo commented on tools/webpack/blocks.js:
```diff
@@ -74,134 +74,209 @@ const createEntrypoints = () => {
+			splitChunks: {
+				cacheGroups: {
+					vendors: {
+						name: 'vendors',
```
How should it be handled when a core block imports a new vendor script that isn't part of the runtime?

DAreRodz replied:
That's a good question. 🤔 I guess we can modify the vendors' bundle's definition so it only includes the common libraries, i.e., `preact`, `@preact/signals`, etc., instead of all dependencies from `node_modules`.

gziolo commented on tools/webpack/blocks.js:
```diff
@@ -74,134 +74,209 @@ const createEntrypoints = () => {
+									process.env.BABEL_CACHE_DIRECTORY || true,
+								babelrc: false,
+								configFile: false,
+								presets: [
```
Does this Babel config respect the setting for `browserslist` that WordPress uses?

DAreRodz replied:
Not sure, but we can change it. I'll take a look.

gziolo commented on tools/webpack/blocks.js:
```diff
@@ -74,134 +74,209 @@ const createEntrypoints = () => {
+		module: {
+			rules: [
+				{
+					test: /\.(j|t)sx?$/,
```
Does it work with TypeScript out of the box now? I think you need to include `@babel/preset-typescript`. See https://github.com/WordPress/gutenberg/pull/28879 for reference.

DAreRodz replied:
Let's leave TypeScript apart for now. I'll remove it from the `test` property.

gziolo replied:
It would work as we don't use TypeScript for core blocks, as far as I know. We can add TS support later.

gziolo commented on lib/client-assets.php:
```diff
@@ -574,3 +574,50 @@ function gutenberg_register_vendor_scripts( $scripts ) {
+		);
+ 	}
+}
+add_action( 'wp_enqueue_scripts', 'gutenberg_register_interactivity_scripts' );
```
I assume the intent is to always register those scripts but enqueue them only when any interactive block is present on the page. In that case `wp_default_scripts` could be the more appropriate hook. In the future, the same script is going to be registered in the WordPress core so a more future-proof strategy would be to override them if present. Some prior art:

https://github.com/WordPress/gutenberg/blob/30e1054c8be9b25ad6723dfc2fd2498c567e574d/lib/client-assets.php#L193-L239

or

https://github.com/WordPress/gutenberg/blob/30e1054c8be9b25ad6723dfc2fd2498c567e574d/lib/client-assets.php#L544-L563

---

Please also note how the paths to the file are being constructed. I also don't think we need the `is_admin()` check. The registration doesn't automatically enqueue scripts. In addition to that, nothing should prevent folks to add a standalone page in WP Admin that renders blocks as if they were on the front end of the site.

DAreRodz replied:
Thanks for the detailed insights. 😊 We can change the hook being used to `wp_default_scripts` and maybe leave the code overriding the action for another PR. Do you agree?

gziolo replied:
`gutenberg_override_script` is a drop-in replacement for `wp_register_script`.

gziolo replied:
I address my comment at https://github.com/WordPress/gutenberg/pull/49994/commits/f4b6c0a9536bf878ba12f54032bcf4186dc3997a. I also moved both filters to the `lib/experiments/interactivity-api` folder. I also think that once we introduce the Interactivity API experiment feature flag, those hooks should only run when the experiment is enabled.

gziolo commented on lib/client-assets.php:
```diff
@@ -574,3 +574,50 @@ function gutenberg_register_vendor_scripts( $scripts ) {
+  * @return string The modified script tag.
+	*/
+function add_defer_attribute( $tag, $handle ) {
+		if ( 0 === strpos( $handle, 'interactivity-' ) ) {
```
I'm curious how it will work later with individual blocks that depend on those scripts. Will they be manually registered with the same prefix? Aside, most of the scripts in the Gutenberg plugin are prefixed with `wp-`.

DAreRodz replied:
Correct, they will be manually registered with the same prefix, at least for now. Maybe it would be a good idea to make the prefix less "generic", like `wp-interactivity-api-`, or similar.

gziolo replied:
Updated to `wp-interactivity-` prefix for now as it is unique enough. See https://github.com/WordPress/gutenberg/pull/49994/commits/6ae760f5e92147285da1cf8dbb8aa44ce8264272.

gziolo commented on packages/block-library/src/utils/interactivity/index.js:
```diff
@@ -0,0 +1,14 @@
+registerDirectives();
+init();
+// eslint-disable-next-line no-console
+console.log( 'Interactivity API started' );
```
Should this printing be guarded to be present only in the dev mode?

DAreRodz replied:
That would be nice. What is the appropriate way to check we are in dev mode in the browser?

luisherranz replied:
This is a related conversation: https://github.com/WordPress/block-interactivity-experiments/pull/222

But in this case, this `console.log` is just a quick way to make sure the Interactivity API loads correctly now that we are integrating everything together. It won't add much value when things "just work", so we can remove it later.

gziolo commented on packages/block-library/src/utils/interactivity/hydration.js:
```diff
@@ -0,0 +1,22 @@
+import { createRootFragment } from './utils';
+import { directivePrefix } from './constants';
+
+export const init = async () => {
```
Is `async` necessary here? There is no usage at the moment that would validate it.

DAreRodz replied:
Actually, yes, the `console.log` in the previous comment should be called after `init` is resolved. I'll change that.

gziolo commented on packages/block-library/package.json:
```diff
@@ -63,12 +65,14 @@
 		"lodash": "^4.17.21",
 		"memize": "^1.1.0",
 		"micromodal": "^0.4.10",
+		"preact": "^10.13.2",
```
Noting that `react` is set as a peer dependency to make integrating with 3rd party projects easier. Would it make sense to follow the same approach for `preact`?

gziolo reviewed (commented):
It's great to see the initial PR opening the Interactivity API in the Gutenberg plugin. I left feedback that concerns integration with the plugin and how it will impact the future development of interactive blocks.

I don't think there is anything blocking it, but we need to have a solid plan for the next steps that ensure that we have the integration of Interactive API with the plugin working seamlessly as we start adding interactive behaviors for individual blocks. My expectation would be that we work on the following task as a follow-up for this PR that makes it easier to use the runtime with PRs that build on top of it:
- moving the existing unit tests for the runtime and directives to the temporary directory `packages/block-library/src/util/interactivity`
- adding JSDoc comments that document the runtime code and directives
- ensuring that the front-end interactivity can be easily wired with core blocks by updating a single place in the codebase if possible
- every behavior added to the block is covered by a corresponding e2e test

gziolo commented on packages/block-library/package.json:
```diff
@@ -27,10 +27,12 @@
 		"src/**/*.scss",
-		"{src,build,build-module}/*/init.js"
+		"{src,build,build-module}/*/init.js",
+		"src/utils/interactivity/index.js"
```
This file is consume through `src` only by React Native, web bundlers would mostly use `build-module`:
```suggestion
		"{src,build,build-module}/utils/interactivity/index.js"
```

luisherranz commented:
Great feedback, thanks Greg 🙂

> ensuring that the front-end interactivity can be easily wired with core blocks by updating a single place in the codebase if possible

I don't think it's worth optimizing that at this moment because it will change once we expose the runtime for experimental use and that shouldn't take long.

gziolo commented:
> > ensuring that the front-end interactivity can be easily wired with core blocks by updating a single place in the codebase if possible
> 
> I don't think it's worth optimizing that at this moment because it will change once we expose the runtime for experimental use and that shouldn't take long.

Fair enough, it should be way easier to draw conclusions once we have 2-3 core blocks working with the runtime 👍🏻 

In fact, there are more comments that I left that might be premature, but now we have it documented we can revisit it later when we make Interactivity API approach the default approach for core blocks.

dmsnell commented:
> This preliminary version is only meant to be used internally by core blocks. We don't even want to expose it yet as experimental.

Fair, but I would hope we would first seek to split these before potentially merging. At least, it's been the case that for many things it's near impossible to extract after the fact, particularly because of the need to design the decoupling into the interfaces.

This is related to the packaging and code organization question, and I know that there's a fairly overwhelming preference in the project to isolate functionality into separate domains. It can still be possible to lock down the code even if it lives in a separate directory.

Just some food for thought.

gziolo committed:
Move and refactor the interactive scritps registration

gziolo committed:
Fix code style violations

gziolo committed:
Use `wp-interactivity-` prefix for script handles

gziolo committed:
Improve the matcher for side effects in `package.json`

gziolo commented:
Here is the list of changes I applied to this PR:
- Rebased it with the latest changes from `trunk`.
- Fixed code style violations with https://github.com/WordPress/gutenberg/pull/49994/commits/3d9447302e1066baddda6c9c5b6841759e7ea301.
- Improved the matcher for side effects in `package.json` of `@wordpress/block-library`.
- Changed the prefix for script handles to `wp-interactivity-` with https://github.com/WordPress/gutenberg/pull/49994/commits/6ae760f5e92147285da1cf8dbb8aa44ce8264272 (@artemiomorales and @SantosGuillamot, this will require some subtle changes in the branches you develop after rebasing).
- I also moved the file and refactored the hooks that register and process interactivity scripts: runtime and vendor with https://github.com/WordPress/gutenberg/pull/49994/commits/f4b6c0a9536bf878ba12f54032bcf4186dc3997a.

Since I applied changes to the branch I haven't started, @DAreRodz and @luisherranz feel free to disagree with any changes I applied and revert or change individual commits.

From my perspective, there are still some follow-up tasks necessary, but I'm open to merging this PR in the current shape whenever we are ready to land one of the open PRs that integrate Interactivity API with the Navigation or the Image block. We could possibly land one of these PRs into this PR, test everything together and merge into `trunk`. Let me know what do you think about that idea.

Some remaining tasks that I want to discuss. They could be as well done as a follow-up(s), not blockers for this PR, but anyone can disagree:
- In https://github.com/WordPress/gutenberg/pull/50041 @SantosGuillamot introduces a feature flag that enables Interactivity API as an experiment. It would be good to put the file that registers the runtime behind the same flag, too.
- webpack config could benefit from further iterations:
    - We might want to list vendors (https://github.com/WordPress/gutenberg/pull/49994#discussion_r1174951301).
    - Ensuring that Babel uses the browser list config from Gutenberg (https://github.com/WordPress/gutenberg/pull/49994#discussion_r1174955539).
    - Ensuring TypeScript integration is fully functional (https://github.com/WordPress/gutenberg/pull/49994#discussion_r1174963151).
    - Deciding whether vendor libraries should be regular or peer dependencies for `@wordpress/block-library` (https://github.com/WordPress/gutenberg/pull/49994#discussion_r1174988136). It isn't important today, but it will be key to do it right once we move Interactivity runtime to its package. It's also what @dmsnell pointed as one of the next steps in https://github.com/WordPress/gutenberg/pull/49994#issuecomment-1518746889.

The other aspects that we need to cover as the implementation matures and we will look into making Interactivity API the default experience - not behind the feature flag would be:
- Moving the runtime and directives to their package.
- Adding JSDoc comments to public methods and other functions that deserve it.
- Bringing and adjusting existing suite of e2e and unit tests from https://github.com/WordPress/block-interactivity-experiments.

I wouldn't rush the second part, as we should focus now on ensuring that everything integrates correctly in the Gutenberg plugin. However, it should be clearly communicated that the code has extensive test coverage present. Maybe it would be valuable to leave a note in the codebase that makes it easy to locate those tests for the time being.

DAreRodz commented:
> Since I applied changes to the branch I haven't started, @DAreRodz and @luisherranz feel free to disagree with any changes I applied and revert or change individual commits.

Thanks for these changes; I think they are OK. 🙂

> We could possibly land one of these PRs into this PR, test everything together and merge into `trunk`. Let me know what do you think about that idea.

It makes sense. 🤔 I believe the Navigation block is almost finished (@SantosGuillamot correct me if I'm wrong). We can use that PR for this purpose.

> * In [Navigation block with the Interactivity API #50041](https://github.com/WordPress/gutenberg/pull/50041) @SantosGuillamot introduces a feature flag that enables Interactivity API as an experiment. It would be good to put the file that registers the runtime behind the same flag, too.

I guess that depends on whether we want a single setting to enable/disable the whole interactivity API or individual settings for each block. Currently, the Interactivity API runtime is only enqueued when a block requires it, so maybe it is not necessary to have a feature flag for this case.


> * We might want to list vendors ([Add Interactivity API runtime #49994 (comment)](https://github.com/WordPress/gutenberg/pull/49994#discussion_r1174951301)).

I did some tests last week, and my idea of listing vendors doesn't quite work. The thing is that dependencies of listed packages are not included in the `vendors` chunk (and we are not going to include those dependencies manually; or maybe it is the only way). I don't know how to solve that problem properly. 🤷

Regarding the rest of the Webpack tasks, I agree they could be done in a follow-up PR. I'll list them in the tracking issue, as well as the other tasks included in what you called the "second part". 🙂

DAreRodz committed:
Add custom useSignalEffect

DAreRodz commented:
I've added our custom `useSignalEffect` to this PR.

spacedmonkey commented on lib/experimental/interactivity-api/script-loader.php:
```diff
@@ -0,0 +1,50 @@
+}
+add_action( 'wp_default_scripts', 'gutenberg_register_interactivity_scripts', 10, 1 );
+
+/**
+ * Adds the "defer" attribute to all the interactivity script tags.
+ *
+ * @param string $tag    The generated script tag.
+ * @param string $handle The script's registered handle.
+ *
+ * @return string The modified script tag.
+ */
+function gutenberg_interactivity_scripts_add_defer_attribute( $tag, $handle ) {
+	if ( 0 === strpos( $handle, 'wp-interactivity-' ) ) {
+		$p = new WP_HTML_Tag_Processor( $tag );
+		$p->next_tag( array( 'tag' => 'script' ) );
+		$p->set_attribute( 'defer', true );
+		return $p->get_updated_html();
+	}
+	return $tag;
+}
+add_filter( 'script_loader_tag', 'gutenberg_interactivity_scripts_add_defer_attribute', 10, 2 );
```
@10upsimon @joemcgill @westonruter Flagging as deferring scripts is going to be part of https://github.com/WordPress/wordpress-develop/pull/4391 / [#12009](https://core.trac.wordpress.org/ticket/12009)

dmsnell replied:
also, have we considered shipping ESModules? I guess there's a compatibility issue with that?

if we _can_, then I think we can instead `$p->set_attribute( 'type', 'module' )` and the proper deferral is automatic.

anyway ignore me if this is wrong or too random.

technically we will want to verify that the `script` tag exists. seems silly in this handler, but I want all Core code to demonstrate proper hygiene

```php
function gutenberg_interactivity_scripts_add_defer_attribute( $tag, $handle ) {
	if ( ! str_starts_with( 'wp-interactivity-', $handle ) ) {
		return $tag;
	}

	$p = new WP_HTML_Tag_Processor( $tag );
	if ( $p->next_tag( 'script' ) ) {
		$p->set_attribute( 'defer', true );
	}

	return $p->get_updated_html();
}
```

dum de dum. thanks for watching!

gziolo replied:
> Flagging as deferring scripts is going to be part of [WordPress/wordpress-develop#4391](https://github.com/WordPress/wordpress-develop/pull/4391) / [#12009](https://core.trac.wordpress.org/ticket/12009)

Thank you for raising awareness, @spacedmonkey. One way to approach it, would be to make all `viewScripts` for the block using the `defer` strategy by default. I briefly discussed it with @felixarntz the other day.

It's worth noting that we will still need this hook for older versions of WordPress core supported by the Gutenberg plugin.

felixarntz replied:
Makes sense, I think once the script loading strategy enhancement has landed in core, we should work on a Gutenberg issue / PR for deferring block scripts by default (🤞 that that is as straightforward as we're thinking).

For the code here, maybe worth adding a `TODO` comment to make this conditionally use the new core API once available.

dmsnell commented on packages/block-library/src/utils/interactivity/hooks.js:
```diff
@@ -0,0 +1,76 @@
+
+	for ( const d in directives ) {
+		const wrapper = directiveMap[ d ]?.( directiveArgs );
+		if ( wrapper !== undefined ) props.children = wrapper;
```
nitpick: we're going to favor _always_ including curlies in any code block. this simple step helps us avoid some really hard-to-discover bugs like the notorious one in Apple's SSL library.

dmsnell commented on packages/block-library/src/utils/interactivity/hooks.js:
```diff
@@ -0,0 +1,76 @@
+		vnode.type = Directive;
+	}
+
+	if ( old ) old( vnode );
```
nitpick

```js
if ( old ) {
	old( vnode );
}
```

gziolo replied:
```js
if ( old ) old( vnode );
```

I think it's automatically formatted this way by Prettier, so you can't do much about it.

gziolo replied:
Update: you can do both, and Prettier will respect it. In the codebase, you can see both ways of representing the line and both are fine for the linter.

dmsnell commented on packages/block-library/src/utils/interactivity/store.js:
```diff
@@ -0,0 +1,45 @@
+	if ( ! storeTag ) return {};
+	try {
+		const { state } = JSON.parse( storeTag.textContent );
+		if ( isObject( state ) ) return state;
```
nitpick: body brackets

dmsnell commented on packages/block-library/src/utils/interactivity/store.js:
```diff
@@ -0,0 +1,45 @@
+	const storeTag = document.querySelector(
+		`script[type="application/json"]#store`
+	);
+	if ( ! storeTag ) return {};
```
nitpick: body brackets

dmsnell commented on packages/block-library/src/utils/interactivity/vdom.js:
```diff
@@ -0,0 +1,94 @@
+	function walk( node ) {
+		const { attributes, nodeType } = node;
+
+		if ( nodeType === 3 ) return [ node.data ];
```
body brackets 🕺

dmsnell commented on packages/block-library/src/utils/interactivity/vdom.js:
```diff
@@ -0,0 +1,94 @@
+			const next = treeWalker.nextSibling();
+			node.remove();
+			return [ null, next ];
+		}
```
the node constants I believe are going to be present _everywhere_ we have a DOM, so why not use them here? I cannot for the life of me remember the difference between `4` and `7`

```js
switch ( nodeType ) {
	case Node.TEXT_NODE:
		return [ node.data ];

	case Node.CDATA_SECTION_NODE: {
		const next = treeWalker.nextSibling();
		node.replaceWith( new window.Text( node.nodeValue ) );
		return [ node.nodeValue, next ];
	}

	case Node.PROCESSING_INSTRUCTION_NODE:
	case Node.COMMENT_NODE: {
		const next = treeWalker.nextSibling();
		node.remove();
		return [ null, next ];	
	}
}
```

the code in here makes more sense to me seeing these

DAreRodz replied:
I agree 😅 I think @luisherranz decided to do so because it saves some bytes to the client, as those props are not mangled.

luisherranz replied:
We can add them as comments.

dmsnell replied:
hm. Are we concerned that much about a few tens of bytes when we're skipping KBs?

> as those props are not mangled

that's the point 🙃
comments at a minimum would be nice. I think we'd be hard pressed to make a case that it matters that we not use the constants that are there 🤷‍♂️

luisherranz replied:
> Are we concerned that much about a few tens of bytes when we're skipping KBs?

Oh, yes. Each byte adds up so if we can get the similar value just by using comments, we should do that 🙂

dmsnell replied:
🤔 🤷‍♂️

dmsnell commented on packages/block-library/src/utils/interactivity/vdom.js:
```diff
@@ -0,0 +1,94 @@
+					__directives: { ignore: true },
+				} ),
+			];
+		if ( island ) hydratedIslands.add( node );
```
🕺

dmsnell commented on packages/block-library/src/utils/interactivity/vdom.js:
```diff
@@ -0,0 +1,94 @@
+			];
+		if ( island ) hydratedIslands.add( node );
+
+		if ( hasDirectives ) props.__directives = directives;
```
🕺

gziolo commented:
> > We could possibly land one of these PRs into this PR, test everything together and merge into `trunk`. Let me know what do you think about that idea.
> 
> It makes sense. 🤔 I believe the Navigation block is almost finished (@SantosGuillamot correct me if I'm wrong). We can use that PR for this purpose.

Yes, that would work. We can also land this PR first, and follow-up with the Navigation block PR. We should do it close to each other to ensure we don't have unused and untestable code in the Gutenberg plugin for too long.
 
> > * In [Navigation block with the Interactivity API #50041](https://github.com/WordPress/gutenberg/pull/50041) @SantosGuillamot introduces a feature flag that enables Interactivity API as an experiment. It would be good to put the file that registers the runtime behind the same flag, too.
> 
> I guess that depends on whether we want a single setting to enable/disable the whole interactivity API or individual settings for each block. Currently, the Interactivity API runtime is only enqueued when a block requires it, so maybe it is not necessary to have a feature flag for this case.

True, we would register these 2 scripts, but they won't be enqueued so maybe it's fine to keep it simple. Regarding the feature flag, it'd be only the Navigation block for now so it's fine as is. We can update it later when we have more blocks ready to be integrated with Interactivity API.
 
> > * We might want to list vendors ([Add Interactivity API runtime #49994 (comment)](https://github.com/WordPress/gutenberg/pull/49994#discussion_r1174951301)).
> 
> I did some tests last week, and my idea of listing vendors doesn't quite work. The thing is that dependencies of listed packages are not included in the `vendors` chunk (and we are not going to include those dependencies manually; or maybe it is the only way). I don't know how to solve that problem properly. 🤷

That is indeed an issue with dependencies of vendor dependencies 🤔 I don't see any simple solution listed in https://webpack.js.org/plugins/split-chunks-plugin/. They have an example with a regular expression for `react` and `react-dom`, but it would probably suffer from the same problem. As long as we don't use any additional vendor libraries in `view.js` files, the current config is good enough. If necessary there is a way to exclude some packages from the list, and that might be a simpler approach.

SantosGuillamot commented:
@gziolo After start using `gutenberg_override_script` in [this commit](https://github.com/WordPress/gutenberg/commit/f4b6c0a9536bf878ba12f54032bcf4186dc3997a), it seems we are sending many JS files to the client. I've been triaging it with @DAreRodz and it seems that they are enqueued in [this function](https://github.com/WordPress/gutenberg/blob/trunk/lib/client-assets.php#L88-L90). Being more specific, it seems to be [this part](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/class-wp-scripts.php#L584-L586) inside the `setTranslations` function. If we comment that code, the extra scripts are not loaded.

Any idea how this could be handled?

_With the current code_

![Screenshot 2023-05-04 at 17 52 15](https://user-images.githubusercontent.com/34552881/236264696-343baa5b-533a-4899-84a1-116b1df91b5f.png)


_Commenting that function or using `wp_register_script`_

![Screenshot 2023-05-04 at 17 52 32](https://user-images.githubusercontent.com/34552881/236264712-3d309336-5539-41a9-b648-8ec1d583c779.png)

DAreRodz committed:
Call `init` after `store` has been initialized

westonruter commented on lib/experimental/interactivity-api/script-loader.php:
```diff
@@ -0,0 +1,50 @@
+ * @return string The modified script tag.
+ */
+function gutenberg_interactivity_scripts_add_defer_attribute( $tag, $handle ) {
+	if ( 0 === strpos( $handle, 'wp-interactivity-' ) ) {
```
```suggestion
	if ( str_starts_with( $handle, 'wp-interactivity-' ) ) {
```

See https://core.trac.wordpress.org/ticket/58012

gziolo replied:
Good catch, thank you @westonruter!

gziolo committed:
Update lib/experimental/interactivity-api/script-loader.php

Co-authored-by: Weston Ruter <westonruter@google.com>

gziolo committed:
Plugin: Ensure that translations are set correctly when overriding scripts
This replicates the same logic in WordPress core:
https://github.com/WordPress/wordpress-develop/blob/a5f3087f6b5d9c52dbe67ed247165dc32427c57d/src/wp-includes/script-loader.php#L306-L308

gziolo commented:
> After start using gutenberg_override_script in [this commit](https://github.com/WordPress/gutenberg/commit/f4b6c0a9536bf878ba12f54032bcf4186dc3997a), it seems we are sending many JS files to the client. I've been triaging it with @DAreRodz and it seems that they are enqueued in [this function](https://github.com/WordPress/gutenberg/blob/trunk/lib/client-assets.php#L88-L90). Being more specific, it seems to be [this part](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/class-wp-scripts.php#L584-L586) inside the setTranslations function. If we comment that code, the extra scripts are not loaded.

I fixed it with https://github.com/WordPress/gutenberg/pull/49994/commits/0cdab0869726d5ebde7f33357456c92fac1b1e2e. I'm not sure why that code was enforcing including `wp-i18n` as a dependency to all scripts coming from the Gutenberg plugin 🤷🏻

I replicated the logic from WordPress core that resolves the issue and ensures that translations still work where applicable:

https://github.com/WordPress/wordpress-develop/blob/a5f3087f6b5d9c52dbe67ed247165dc32427c57d/src/wp-includes/script-loader.php#L306-L308

I also rebased the branch to include the fix that resolves the issue with the Navigation block not working correctly on the front end when using the version from `trunk`.

The post editor when using Polish translations (I had to update translation using the admin interface to see them correctly applied):

<img width="1245" alt="Screenshot 2023-05-05 at 11 37 44" src="https://user-images.githubusercontent.com/699132/236424965-55611cff-f512-47c1-ad20-0b6139a9281f.png">

gziolo reviewed (approved):
I think this PR is now in good shape to land in the `trunk`. I summarized the latest status in https://github.com/WordPress/gutenberg/pull/49994#issuecomment-1527560984 and followed-up with https://github.com/WordPress/gutenberg/pull/49994#issuecomment-1534568431.

There are still some comments to address left by @felixarntz and @dmsnell that it would be great to address before merging it. I applied a couple of commits to this branch, so I would appreciate some cross-checking from someone else. I'm mostly expressing that I don't see any blockers for this PR, but there is still follow-up work that can happen while we expose the Interactivity API behind the feature flag setting to the users.

DAreRodz commented:
Thanks for the review, @gziolo 😊  I've been testing the latest changes with @SantosGuillamot using the Navigation block on https://github.com/WordPress/gutenberg/pull/50041, and everything seems to be working now.
